File: frontend/src/api/bucketPurge.ts
/*
 * Copyright (c) 2026 Laurent Barbe
 * Licensed under the Apache License, Version 2.0
 */
import type { S3AccountSelector } from "./accountParams";
import type { BucketOperationTarget } from "./bucketOperation";
import {
  buildJsonPostRequestInit,
  resolveApiBaseUrl,
  streamBucketsWithSse,
} from "./sseBucketsStream";

export type BucketPurgePayload = {
  buckets?: string[];
  targets?: BucketOperationTarget[];
  parallelism?: number;
  include_versions?: boolean;
  confirmation: string;
};

export type BucketDeleteWithPurgePayload = {
  parallelism?: number;
  confirmation: string;
};

export type BucketPurgeFailure = {
  bucket_name: string;
  key?: string | null;
  version_id?: string | null;
  stage: "list" | "delete" | "versions" | "delete_bucket";
  message: string;
  count: number;
};

export type BucketPurgeBucketResult = {
  bucket_name: string;
  context_id?: string | null;
  context_name?: string | null;
  status: "completed" | "completed_with_errors" | "failed" | "canceled";
  listed_objects: number;
  listed_versions: number;
  deleted_objects: number;
  deleted_versions: number;
  failed_count: number;
  bucket_deleted: boolean;
  duration_seconds: number;
  failures_sample: BucketPurgeFailure[];
};

export type BucketPurgeProgress = {
  request_id?: string | null;
  stage: "prepare" | "list" | "delete" | "versions" | "delete_bucket" | "completed";
  bucket_name?: string | null;
  context_id?: string | null;
  context_name?: string | null;
  total_buckets: number;
  completed_buckets: number;
  listed_objects: number;
  listed_versions: number;
  deleted_objects: number;
  deleted_versions: number;
  total_entries_estimate?: number | null;
  total_entries_final?: boolean;
  failed_count: number;
  bucket_deleted: boolean;
  message?: string | null;
};

export type BucketPurgeResult = {
  status: "completed" | "completed_with_errors" | "failed" | "canceled";
  total_buckets: number;
  completed_buckets: number;
  listed_objects: number;
  listed_versions: number;
  deleted_objects: number;
  deleted_versions: number;
  failed_count: number;
  bucket_deleted: boolean;
  started_at: string;
  finished_at: string;
  buckets: BucketPurgeBucketResult[];
};

type BucketPurgeStreamOptions = {
  signal?: AbortSignal;
  onProgress?: (event: BucketPurgeProgress) => void;
};

export function streamManagerBucketPurge(
  contextId: S3AccountSelector,
  payload: BucketPurgePayload,
  options?: BucketPurgeStreamOptions
): Promise<BucketPurgeResult> {
  const baseUrl = resolveApiBaseUrl();
  const query = new URLSearchParams({ account_id: String(contextId) });
  return streamBucketsWithSse<BucketPurgeProgress, BucketPurgeResult>({
    url: `${baseUrl}/manager/bucket-purge/stream?${query.toString()}`,
    options,
    requestInit: buildJsonPostRequestInit(payload),
    streamFailedLabel: "Bucket purge stream failed",
    missingResultMessage: "Bucket purge stream ended without a result payload",
  });
}

export function streamManagerBucketDeleteWithPurge(
  contextId: S3AccountSelector,
  bucketName: string,
  payload: BucketDeleteWithPurgePayload,
  options?: BucketPurgeStreamOptions
): Promise<BucketPurgeResult> {
  const baseUrl = resolveApiBaseUrl();
  const query = new URLSearchParams({ account_id: String(contextId) });
  return streamBucketsWithSse<BucketPurgeProgress, BucketPurgeResult>({
    url: `${baseUrl}/manager/buckets/${encodeURIComponent(bucketName)}/delete/stream?${query.toString()}`,
    options,
    requestInit: buildJsonPostRequestInit(payload),
    streamFailedLabel: "Bucket deletion stream failed",
    missingResultMessage: "Bucket deletion stream ended without a result payload",
  });
}

export function streamCephAdminBucketPurge(
  endpointId: number,
  payload: BucketPurgePayload,
  options?: BucketPurgeStreamOptions
): Promise<BucketPurgeResult> {
  const baseUrl = resolveApiBaseUrl();
  return streamBucketsWithSse<BucketPurgeProgress, BucketPurgeResult>({
    url: `${baseUrl}/ceph-admin/endpoints/${endpointId}/buckets/purge/stream`,
    options,
    requestInit: buildJsonPostRequestInit(payload),
    streamFailedLabel: "Bucket purge stream failed",
    missingResultMessage: "Bucket purge stream ended without a result payload",
  });
}

export function streamStorageOpsBucketPurge(
  payload: BucketPurgePayload,
  options?: BucketPurgeStreamOptions
): Promise<BucketPurgeResult> {
  const baseUrl = resolveApiBaseUrl();
  return streamBucketsWithSse<BucketPurgeProgress, BucketPurgeResult>({
    url: `${baseUrl}/storage-ops/buckets/purge/stream`,
    options,
    requestInit: buildJsonPostRequestInit(payload),
    streamFailedLabel: "Bucket purge stream failed",
    missingResultMessage: "Bucket purge stream ended without a result payload",
  });
}

File: backend/app/models/bucket_purge.py
# Copyright (c) 2026 Laurent Barbe
# Licensed under the Apache License, Version 2.0
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import Field

from app.models.base import ApiModel
from app.models.bucket_operation import (
    BucketOperationRequest,
    BucketOperationTarget,
)


BucketPurgeStatus = Literal["completed", "completed_with_errors", "failed", "canceled"]
BucketPurgeStage = Literal["prepare", "list", "delete", "versions", "delete_bucket", "completed"]


class BucketPurgeRequest(BucketOperationRequest):
    targets: list[BucketOperationTarget] = Field(default_factory=list, max_length=200)
    parallelism: int = Field(default=10, ge=1, le=64)
    include_versions: bool = True
    confirmation: str = ""

def bucket_purge_confirmation_phrase(target_count: int) -> str:
    return f"PURGE {target_count} BUCKETS"


class BucketDeleteWithPurgeRequest(ApiModel):
    parallelism: int = Field(default=10, ge=1, le=64)
    confirmation: str = ""


def bucket_delete_with_purge_confirmation_phrase(bucket_name: str) -> str:
    return f"DELETE BUCKET {bucket_name}"


class BucketPurgeFailure(ApiModel):
    bucket_name: str
    stage: str
    message: str
    key: Optional[str] = None
    version_id: Optional[str] = None
    count: int = 0


class BucketPurgeBucketResult(ApiModel):
    bucket_name: str
    context_id: Optional[str] = None
    context_name: Optional[str] = None
    status: BucketPurgeStatus
    listed_objects: int = 0
    listed_versions: int = 0
    deleted_objects: int = 0
    deleted_versions: int = 0
    failed_count: int = 0
    bucket_deleted: bool = False
    duration_seconds: float = 0
    failures_sample: list[BucketPurgeFailure] = Field(default_factory=list)


class BucketPurgeProgress(ApiModel):
    request_id: Optional[str] = None
    stage: BucketPurgeStage = "prepare"
    bucket_name: Optional[str] = None
    context_id: Optional[str] = None
    context_name: Optional[str] = None
    total_buckets: int = 0
    completed_buckets: int = 0
    listed_objects: int = 0
    listed_versions: int = 0
    deleted_objects: int = 0
    deleted_versions: int = 0
    total_entries_estimate: Optional[int] = None
    total_entries_final: bool = False
    failed_count: int = 0
    bucket_deleted: bool = False
    message: Optional[str] = None


class BucketPurgeResult(ApiModel):
    status: BucketPurgeStatus
    total_buckets: int = 0
    completed_buckets: int = 0
    listed_objects: int = 0
    listed_versions: int = 0
    deleted_objects: int = 0
    deleted_versions: int = 0
    failed_count: int = 0
    bucket_deleted: bool = False
    started_at: datetime
    finished_at: datetime
    buckets: list[BucketPurgeBucketResult] = Field(default_factory=list)
